docs/geps/gep-01.rst

```diff
@@ -0,0 +1,139 @@
+.. _gep-1:
+
+==========================
+GEP 1 — Naming Conventions
+==========================
+
+:Author: `Eric Sommer <https://github.com/Eric-Sommer>`_
+:Author: `Hans-Martin von Gaudecker <https://github.com/hmgaudecker>`_
+:Status: Draft
+:Type: Standards Track
+:Created: 2019-11-04
+:Resolution: <url> (required for Accepted | Rejected | Withdrawn)
+
+
+Abstract
+--------
+
+This GEP pins down naming conventions for GETTSIM — general rules for what data columns,
+parameters, Python identifiers (functions, variables), etc. should be called. In a
+nutshell and without explanations, these conventions are:
+
+* Names follow standard Python conventions (``lowercase_with_underscores``)
+* Names should be long enough to be readable, but
+
+  - For column names in the user-facing API, there is a hard limit of 15 characters
+  - For others, there is a soft limit of 15 and a hard limit of 20 characters
+
+* The language should generally be German
```

Comment:
I agree that gettsim-internal (tax-transfer-related) variables should be named in German. But should this also hold for input and output data? Do we want `alter` instead of `age` and `familienstand` instead of `marital_status`?

Reply:
For the reference implementation, I think so, yes. There is plenty of German already and some of it would be hard to translate. Also, it would be useful to have stuff like `gkv` line up with `gkv_beitrag` rather than trying to figure out from the English name that something is related to the same concept.

It is trivial to provide an English translation, so this mostly affects the internal workings of the code.

Reply:
I also think it would feel strange to work with only German variables and functions. Additionally, it would make it almost impossible for non-German speakers to participate in the project.

We have this nice differentiation in three basic building blocks of the code:
>1.     The input and output data, including  any values stored intermediately. The scope of this GEP are the column identifiers. The index and the variable contents are defined elsewhere could link if we have something).
> 2.     The parameters of the tax transfer system, as detailed in the YAML files.
 >3.     Python identifiers, that is, variables and functions.

So I am more in favor of having 1. and 3. in English and only 2. in German. 

> There is plenty of German already and some of it would be hard to translate.
That's a good point. What do you those guys, who have worked more with the code think? In any case we should decide for one language and change all variables, functions (and comments) accordingly.

Reply:
> I also think it would feel very strange to work with only German variables and functions. @hmgaudecker, would you then have also docstrings and comments in German?

No, only identifiers. 

> Additionally, it would make it almost impossible for non-German speakers to participate in the project.
> 
> We have this nice differentiation in three basic building blocks of the code:
> 
> > 1. The input and output data, including  any values stored intermediately. The scope of this GEP are the column identifiers. The index and the variable contents are defined elsewhere could link if we have something).

I would say no with respect to the scope of this GEP. I do not think it makes sense to separate these three things at the level of pinning down naming conventions.

> > 2. The parameters of the tax transfer system, as detailed in the YAML files.

> > 3. Python identifiers, that is, variables and functions.

> So I am more in favor of having 1. and 3. in English and only 2. in German.
> 
> > There is plenty of German already and some of it would be hard to translate.
> 
> That's a good point. In any case we should decide for one language and change all variables, functions (and comments) accordingly. What do the guys, who have worked more with the code, think is easier?

I was not trying to suggest that and I think it would be a bad idea indeed. Sorry if anything gave that impression.

> > Also, it would be useful to have stuff like gkv line up with gkv_beitrag rather than trying to figure out from the English name that something is related to the same concept.
> 
> I think it would be an option to have all names that are directly related to an institutional feature (parameter in tax-transfer-system) in German. If we adopt this proposal, we could have a function name like `calculate_elterngeld()`. I prefer this over `berechne_elterngeld()`

I am totally in favour of that. Maybe something like "verbs should always be in English"? 

We will continue to have a mix of languages in 3. (Python identifiers) for sure. 

But I personally would prefer to get rid of the language mixing in the column identifiers and tax / transfer parameters. There are too many grey areas what constitutes an institutional feature and what does not. Some components of income / deductions might have reasonable translations, but "werbungskosten" or "zu_versteuerndes_einkommen" ? I would then prefer having only German names in the formula. Same for other stuff. "Eheähnliche Gemeinschaft" ?

Reply:
Ok, I misunderstood something. We agree on 2. (German) and 3. (English - except referring directly to something institutional). The open question is block 1. 

I tend to agree with your proposed solution then (German).

Reply:
Upon re-reading, I see where the confusion is coming from. Sorry about that! It is a draft, but still...

Reply:
Upon re-reading, I see where the confusion is coming from. Sorry about that! It is a draft, but still...

Reply:
> I do not think it makes sense to separate these three things at the level of pinning down naming conventions.

I disagree. Once the household API is set up, we will require relatively few input columns (30-40) which are generic enough to abstract from German tax-transfer specifics. `marital_status` might be one exception but we'd need to provide some additional guidelines (data type, scope, ...) anyway. As @ChristianZimpelmann, demanding `alter`, `weiblich` and `am_status` would raise additional burden for non-German users and possibly for German ones as well.

Reply:
Again, I was not precise enough. I tried to say "this should be in one GEP rather than in three", not that rules for all three broad concepts need to be the same.

And again: the question is not about the API (trivial to provide in various formats), but about the internal code. And there we will end up with all those cases (health insurance, ...) where there is a 1:1 link between the parameters and the data columns. 

As for `weiblich`: To be future-proof we should go for `geschlecht` anyhow to allow for more than two values ...

In the end of the day, I think it will be hard or unnatural to avoid German altogether there and I do not see a good reason to do so in the forest place. I tend to prefer going all the way in terms of German column names; but if we decide in the end of the day that we should leave stuff like `age` in English, this would be fine, too, of course.